### PR TITLE
Adds Jedis constructor, general cleanup

### PR DESCRIPTION
**Pull request by bsbodden**
* Adds Jedis instance constructor + tests
* Removes all unused imports
* Fixed all deprecation warning
* Updates JUnit 4 to latest release

### CODE REVIEW

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Please replace the tab characters with spaces.

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraph.java [hunk 1]
```diff
@@ -120,9 +126,14 @@ protected ResultSet sendReadOnlyQuery(String graphId, String preparedQuery, long
...
-        this.client.close();
-    }
+	public void close() {
+		if (pool != null) {
+			pool.close();
+		}
+		if (jedis != null) {
+			jedis.close();
+		}
+	}
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraph.java:136]:
```suggestion
    public void close() {
        if (pool != null) {
            pool.close();
        }
        if (jedis != null) {
            jedis.close();
        }
    }
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 2]
```diff
@@ -38,7 +38,8 @@ public RedisGraphPipeline(Client client, RedisGraph redisGraph){
     public Response<ResultSet> query(String graphId, String query) {
         client.sendCommand(RedisGraphCommand.QUERY, graphId, query, Utils.COMPACT_STRING);
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:42]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 3]
```diff
@@ -55,7 +56,8 @@ public ResultSet build(Object o) {
     public Response<ResultSet> readOnlyQuery(String graphId, String query) {
         client.sendCommand(RedisGraphCommand.RO_QUERY, graphId, query, Utils.COMPACT_STRING);
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:60]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 4]
```diff
@@ -75,7 +77,8 @@ public ResultSet build(Object o) {
     public Response<ResultSet> query(String graphId, String query, long timeout) {
         client.sendCommand(RedisGraphCommand.QUERY, graphId, query, Utils.COMPACT_STRING, Utils.TIMEOUT_STRING, Long.toString(timeout));
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:81]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 5]
```diff
@@ -95,7 +98,8 @@ public ResultSet build(Object o) {
     public Response<ResultSet> readOnlyQuery(String graphId, String query, long timeout) {
         client.sendCommand(RedisGraphCommand.RO_QUERY, graphId, query, Utils.COMPACT_STRING, Utils.TIMEOUT_STRING, Long.toString(timeout));
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:102]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 6]
```diff
@@ -114,7 +118,8 @@ public ResultSet build(Object o) {
         String preparedQuery = Utils.prepareQuery(query, params);
         client.sendCommand(RedisGraphCommand.QUERY, graphId, preparedQuery, Utils.COMPACT_STRING);
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:122]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 7]
```diff
@@ -133,7 +138,8 @@ public ResultSet build(Object o) {
         String preparedQuery = Utils.prepareQuery(query, params);
         client.sendCommand(RedisGraphCommand.RO_QUERY, graphId, preparedQuery, Utils.COMPACT_STRING);
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:142]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 8]
```diff
@@ -156,7 +162,8 @@ public ResultSet build(Object o) {
         String preparedQuery = Utils.prepareQuery(query, params);
         client.sendCommand(RedisGraphCommand.QUERY, graphId, preparedQuery, Utils.COMPACT_STRING, Utils.TIMEOUT_STRING, Long.toString(timeout));
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:166]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 9]
```diff
@@ -179,7 +186,8 @@ public ResultSet build(Object o) {
         String preparedQuery = Utils.prepareQuery(query, params);
         client.sendCommand(RedisGraphCommand.RO_QUERY, graphId, preparedQuery, Utils.COMPACT_STRING, Utils.TIMEOUT_STRING, Long.toString(timeout));
         return getResponse(new Builder<ResultSet>() {
-            @Override
+            @SuppressWarnings("unchecked")
+			@Override
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:190]:
```suggestion
            @SuppressWarnings("unchecked")
            @Override
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 10]
```diff
@@ -29,26 +31,18 @@ public void deleteGraph() {
...
     public void testSyntaxErrorReporting() {
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Type mismatch: expected String but was Integer");
-
-        // Issue a query that causes a compile-time error
-        api.query("social", "RETURN toUpper(5)");
-
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> api.query("social", "RETURN toUpper(5)"));
+            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:38]:
```suggestion
        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> api.query("social", "RETURN toUpper(5)"));
            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 11]
```diff
@@ -29,26 +31,18 @@ public void deleteGraph() {
...
     @Test
     public void testRuntimeErrorReporting() {
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Type mismatch: expected String but was Integer");
-
-        // Issue a query that causes a run-time error
-        api.query("social", "MATCH (p:person) RETURN toUpper(p.mixed_prop)");
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> api.query("social", "MATCH (p:person) RETURN toUpper(p.mixed_prop)"));
+            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:45]:
```suggestion
        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> api.query("social", "MATCH (p:person) RETURN toUpper(p.mixed_prop)"));
            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 12]
```diff
@@ -79,37 +73,34 @@ public void testExceptionFlow() {
...
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Type mismatch: expected String but was Integer");
         RedisGraphContext c = api.getContext();
 
-        // Issue a query that causes a compile-time error
-        c.query("social", "RETURN toUpper(5)");
-
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> c.query("social", "RETURN toUpper(5)"));
+            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:80]:
```suggestion
        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> c.query("social", "RETURN toUpper(5)"));
            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 13]
```diff
@@ -79,37 +73,34 @@ public void testExceptionFlow() {
...
     }
 
     @Test
     public void testMissingParametersSyntaxErrorReporting(){
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Missing parameters");
-        api.query("social","RETURN $param");
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> api.query("social", "RETURN $param"));
+            assertTrue(exception.getMessage().contains("Missing parameters"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:87]:
```suggestion
        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> api.query("social", "RETURN $param"));
            assertTrue(exception.getMessage().contains("Missing parameters"));
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 14]
```diff
@@ -79,37 +73,34 @@ public void testExceptionFlow() {
...
     }
 
     @Test
     public void testMissingParametersSyntaxErrorReporting2(){
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Missing parameters");
-        api.query("social","RETURN $param", new HashMap<>());
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> api.query("social", "RETURN $param", new HashMap<>()));
+            assertTrue(exception.getMessage().contains("Missing parameters"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:94]:
```suggestion
        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> api.query("social", "RETURN $param", new HashMap<>()));
            assertTrue(exception.getMessage().contains("Missing parameters"));
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 15]
```diff
@@ -79,37 +73,34 @@ public void testExceptionFlow() {
...
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Type mismatch: expected String but was Integer");
-
         RedisGraphContext c = api.getContext();
-        // Issue a query that causes a run-time error
-        c.query("social", "MATCH (p:person) RETURN toUpper(p.mixed_prop)");
+
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> c.query("social", "MATCH (p:person) RETURN toUpper(p.mixed_prop)"));
+            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:103]:
```suggestion

        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> c.query("social", "MATCH (p:person) RETURN toUpper(p.mixed_prop)"));
            assertTrue(exception.getMessage().contains("Type mismatch: expected String but was Integer"));
```

File: src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java [hunk 16]
```diff
@@ -140,10 +131,9 @@ public void testContextExceptionFlow() {
...
     @Test
-    public void timeoutExcpetion() {
-        exceptionRule.expect(JRedisGraphException.class);
-        exceptionRule.expectMessage("Query timed out");
-
-        api.query("social", "UNWIND range(0,100000) AS x WITH x AS x WHERE x = 10000 RETURN x", 1L);
+    public void timeoutException() {
+    	JRedisGraphException exception = assertThrows(JRedisGraphException.class,
+                () -> api.query("social", "UNWIND range(0,100000) AS x WITH x AS x WHERE x = 10000 RETURN x", 1L));
+            assertTrue(exception.getMessage().contains("Query timed out"));
```
Comment (sazzad16) [inline src/test/java/com/redislabs/redisgraph/exceptions/JRedisGraphErrorTest.java:137]:
```suggestion
    public void timeoutException() {
        JRedisGraphException exception = assertThrows(JRedisGraphException.class,
                () -> api.query("social", "UNWIND range(0,100000) AS x WITH x AS x WHERE x = 10000 RETURN x", 1L));
            assertTrue(exception.getMessage().contains("Query timed out"));
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraph.java [hunk 17]
```diff
@@ -32,26 +33,31 @@ public RedisGraph() {
...
-
 
     /**
      * Overrides the abstract function. Gets and returns a Jedis connection from the Jedis pool
      * @return a Jedis connection
      */
     @Override
     protected Jedis getConnection() {
-        return client.getResource();
+      return jedis != null ? jedis : pool.getResource();
```
Comment (sazzad16) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraph.java:60]:
```suggestion
        return jedis != null ? jedis : pool.getResource();
```

**Comment by bsbodden** [comment]
> Please replace the tab characters with spaces.

Done. Thanks.

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**